ui-admin/src/util/RequireUserPermission.tsx

```diff
@@ -0,0 +1,55 @@
+import React from 'react'
+import {
+  userHasPermission,
+  useUser
+} from 'user/UserProvider'
+import { Portal } from '@juniper/ui-core'
+
+
+export type RequireUserPermissionProps = RequireSuperuser | RequirePermissions
+
+type BaseRequireUserPermissionProps = {
+  noPermissionMessage?: React.ReactNode,
+  children: React.ReactNode,
+}
+
+type RequireSuperuser = BaseRequireUserPermissionProps & {
+  superuser: true
+}
+
+const isRequireSuperuser = (props: RequireUserPermissionProps): props is RequireSuperuser => {
+  return (props as RequireSuperuser).superuser
+}
+
+type RequirePermissions = BaseRequireUserPermissionProps &  {
+  portal: Portal,
+  perms: string[],
+}
+
+const isRequirePermissions = (props: RequireUserPermissionProps): props is RequirePermissions => {
+  return (props as RequirePermissions).perms !== undefined
+}
+
+export const RequireUserPermission = (props: RequireUserPermissionProps) => {
+  const { user } = useUser()
+
+  const hasPermission = () => {
+    if (isRequireSuperuser(props)) {
+      return user?.superuser
+    }
+    if (isRequirePermissions(props)) {
+      return props.perms.every(perm => userHasPermission(user, props.portal.id, perm))
+    }
+    return false
+  }
+
+  if (hasPermission()) {
+    return <>
```

Comment:
do you need to wrap this in a fragment?  can you just `return props.children`?

Reply:
surprisingly, yes. there might be a type other than `React.ReactNode` that plays nicer, but it technically isn't an element, so it will complain that `string | null | <etc...> is not a valid react element` unless it's mounted in JSX. So the fragment is necessary.